microprofile/WFLY-12590_Integrate_MicroProfile_Fault_Tolerance_2.0.adoc

```diff
@@ -0,0 +1,99 @@
+This proposal pertains to MP FT version 2.0 given the timing of this RFE.
+The integration is based on SmallRye project, specifically `smallrye-fault-tolerance` module.
+The `smallrye-fault-tolerance` module is already in use by projects https://thorntail.io/[Thorntail] (WildFly Swarm) and https://quarkus.io/[Quarkus].
+
+MicroProfile Fault Tolerance specification simplifies the challenges brought by the distributed nature of microservices.
+This is because the communication with external systems is inherently unreliable.
+This increases demand on resiliency of applications.
+
+The patterns brought by this specification are annotations such as `@Timeout`, `@Fallback`, `@Retry`, `@CircuitBreaker` and `@Bulkhead`.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-12590[WFLY-12590 Integrate MicroProfile Fault Tolerance 2.0]
+* https://issues.jboss.org/browse/EAP7-1347[EAP7-1347 Integrate MP Fault Tolerance 2.0 into EAP]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/WFLY-10522[WFLY-10522 Support Eclipse MicroProfile Config] - dependency; already integrated
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:mnovak@redhat.com[Miroslav Novak]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE.
+// Discuss with QE during the Kickoff state to decide this
+[ ] Engineering
+
+[x] QE
+
+=== Affected Projects or Components
+
+* The specification repository resides at official Eclipse organization on GitHub in https://github.com/eclipse/microprofile-fault-tolerance[eclipse/microprofile-fault-tolerance] repository.
+* The implementation of Eclipse MicroProfile Fault Tolerance will be provided by SmallRye project https://github.com/smallrye/smallrye-fault-tolerance[smallrye/smallrye-fault-tolerance].
+* The underlying library used by _current_ SmallRye implementation is Hystrix which resides in official Netflix GitHub organization https://github.com/Netflix/Hystrix[Netflix/Hystrix].
+
+=== Other Interested Projects
+
+Not interested in our integration per se, but projects that have already integrated MP FT:
+
+* https://thorntail.io/[Thorntail] has integrated MicroProfile Fault Tolerance as a https://github.com/thorntail/thorntail/tree/master/fractions/microprofile/microprofile-fault-tolerance[fraction].
+* https://quarkus.io/[Quarkus] has integrated MicroProfile Fault Tolerance as an https://github.com/quarkusio/quarkus/tree/master/extensions/smallrye-fault-tolerance[extension].
+
+== Requirements
+
+* Support Eclipse MicroProfile Fault Tolerance version 2.0 for all applications deployed in WildFly.
+* [TO BE CLARIFIED] The extension is either added to all configurations provided by WildFly (standalone and domain mode)
+  or provided as a layer in a separate build profile.
+* Upstream Eclipse MicroProfile Fault Tolerance TCK test suite will be added to `microprofile-tck` WildFly upstream test suite.
+
+== Nice-to-Have Requirements
```

Comment:
@rhusar it would be great to provide some information about state of CircuitBreaker/Bulkhead using CLI operation on the deployment. For example it would list all fault tolerance annotations on methods and provide information about its status and configured values.

For example whether `@CircuitBreaker` is open/closed/semi-open. For `@Bulkhead` how many requests is in progress and how many in `waitingTaskQueue`.

CLI operation which would list all open CircuitBreakers might be handle as well.

Looking at https://download.eclipse.org/microprofile/microprofile-fault-tolerance-2.0/microprofile-fault-tolerance-spec.html#_integration_with_microprofile_metrics - this is for free when MP Metrics is added. Could it be enabled on the fly without having to redeploy application?

Reply:
This would be great to have in Web Console as well - cc-ing @hpehl wdyt?

Reply:
Sounds like a good idea. If there are management operations available, bringing the results into the console makes sense to me.

AIUI the management operations are currently not part of this RFE, right?

Reply:
@hpehl No RFE exist yet, I think we need to discuss with @rhusar whether this will be follow up RFE or it will part of this one.

Reply:
Currently, I think it's unlikely that we will be exposing management resources or operations for any of this (or any other MP specs). The MP Metrics will be the only interface for this since it's the best fit.

Reply:
@rhusar Looking at other MP specs in WF there are exposed some management resources, for example for Config, Health or Metrics. It sounds complicated to query metrics rest api for those. Could you raise this question on Monday's Microprofile meeting what is general strategy, please?

Reply:
@rhusar any update from the meeting?

Reply:
> @rhusar Looking at other MP specs in WF there are exposed some management resources, for example for Config, Health or Metrics.

Those are configuration resources, none of those is exposing any runtime resources/ops.

Reply:
> @rhusar any update from the meeting?

The gist is we want to avoid exposing anything extra deployment resources/ops wherever possible. In case we decide to expose anything that we deem necessary (needs to be investigated) this would probably be in the next iteration of the development. 

What we will be currently exposing is as I mentioned earlier in metrics: https://download.eclipse.org/microprofile/microprofile-fault-tolerance-2.0/microprofile-fault-tolerance-spec.html#_integration_with_microprofile_metrics